### PR TITLE
Stats over time

### LINKED ISSUE #188: View participation stats over a time range
### Overview:
Add an ability for admins to view statistics over a specified time range

### Details:
- Create a page called Participation Statistics
- On the page there should be two date selectors and two time selectors (or one calendar that allows you to select start/end dates
- Once the date is selected, it should query all attendance objects from start date/time to stop date/time.
- The default start date is Jan 1st of the current year. The default end date is today
- Default start time should be 12:00 AM. The default end time is the current time
- The UI should display the following statistics:

> - [ ] Total Events (unique events ids)
> - [ ] Total Attendance (the sum of the number of volunteers for all events. If a volunteer checked in and out multiple times in a single event, we should still count it as once)
> - [ ] Total hours worked (sum of all attendance hour diffs)

You can view this [figma](https://www.figma.com/file/ETUKCDm2THyAdUsRpWlPlM/Helping-Mamas-%2F-Spr23?node-id=2675%3A28708&t=CwnQr3QRR8IKVKVz-1)

### LINKED ISSUE #193: On participation statistics, add date/time filter
Users should be able to filter for attendance objects within two given date/times on their personal participation history page

Provide sum for those objects within the range.

This should have the same design as the view attendance objects here #188

### PR DESCRIPTION
**Pull request by ShmulTomer**
Making progress but still have work to do on this (I can probably have it done by late tonight)

### CODE REVIEW

**Comment by jwu6492** [comment]
@ShmulTomer you might be able to reference my changes on `StatDisplay.jsx` and see if that can help you

**Comment by YiQin93** [comment]
I loved the date range selector!! It looks very elegant!! A medal for all of you 🥇 :)

**About the event history table.** 
It has the following issues:

1. It would be great if we could add a title for it, named "Event History" and be left-aligned at the top of the table.

2. For the table's header:
- Expected: "Event Name", "Date", "Time", and "Attendance"
- What it has: the current header is for the Volunteers database/list.

3. For the content on the table
- Expected: the event's name, date, time span, the number of volunteers who participated, and a "View Event Page" button.
- What it has: the event's name, date & time but in a hard-to-read format.

Since it looks like it will take some time to make these changes, I would like your opinions. We can either:

A. Fix all three points above and make it look like the design.

B. Remove the Event History table for this deployment, but keep the code, and we will fix it next semester.

This table is a nice-to-have feature. It makes it easy for users to review previous events over a period of time, as well as understand which event contributes the most/least hours of work. It would work best with everything arranged and labeled. 

Therefore, I would love to know your capacity in this. If you do not have sufficient time or just do not want to work on it now, we will choose option B!

Let me know!

**Another small thing:**
It would be great if we could left-aligned the header "Statistics". It is currently at the center. Keeping the same alignment makes it easier for the users to find information!

**Comment by ShmulTomer** [comment]
Hey Yi! Thank you for the feedback! I'm currently on a blocker for the backend of this screen, that if I can fix today then also finishing the front end will be no big deal. I'm keeping up with Colin on this as well

File: src/pages/api/attendance/eventstatistics.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        
+            }
+        }
+
+    ]);
+    if (!attendance) {
+        return res.status(400);
+      }
+        
+    // var groupBy = function(xs, key) {
```
Comment (colin4554) [inline src/pages/api/attendance/eventstatistics.ts]: Please delete any comments we don't need

File: src/actions/queries.js [hunk 2]
```diff
@@ -133,6 +138,16 @@ export const updateEventById = (id, event) =>
 export const getAttendanceForEvent = (eventId) =>
   axios.post("/api/attendance/statistics", { eventId });
 
+export const getSimpleAttendanceForEvent = (eventId) =>
+  axios.post("/api/attendance/simplestatistics", { eventId });
+
+//export const getEventStatistics = (startDate, endDate) =>
+export const getEventStatistics = () =>
```
Comment (colin4554) [inline src/actions/queries.js]: Why does this not take a start and end date anymore? On the mongo side, you should be able to easily filter the attendance objects by checkIn with something like: `{checkedIm:{$gte:startDate,$lt:endDate}`

File: src/actions/queries.js [hunk 3]
```diff
@@ -133,6 +138,16 @@ export const updateEventById = (id, event) =>
 export const getAttendanceForEvent = (eventId) =>
   axios.post("/api/attendance/statistics", { eventId });
 
+export const getSimpleAttendanceForEvent = (eventId) =>
```
Comment (colin4554) [inline src/actions/queries.js]: Please remove any functions we aren't using (on both the front and back end)

File: src/pages/api/users/stats.js [hunk 4]
```diff
@@ -1,13 +1,18 @@
...
+  getAttendanceByUserID,
 } = require("../../../../server/actions/volunteerstats");
 
 export default async function handler(req, res, next) {
   if (req.method === "GET") {
-    let event = await getEventsByUserID(req.query.volunteer, next);
+    let attendances;
+    if (req.query.volunteer != "null")
+      attendances = await getAttendanceByUserID(req.query.volunteer, next);
+    else attendances = await getAllAttendanceObjects(next);
```
Comment (colin4554) [inline src/pages/api/users/stats.js]: Not sure if this was you or Jonathan, but we should just return an empty list if a userId isn't passed

**Review by colin4554 — CHANGES_REQUESTED** [review]
Let me know if you have any questions

**Comment by YiQin93** [comment]
Overall it looks fantastic! I am so glad that this finally worked out!!!

There seem to be some issues with the calculation of participated hours. For example,
- In steps [#94](https://go.playerzero.app/share/63decbb12b1d69c0be54997a/devtools/94/Comments) to [#97](https://go.playerzero.app/share/63decbb12b1d69c0be54997a/devtools/97/Comments) in this report: (https://go.playerzero.app/share/63decbb12b1d69c0be54997a), there is a Yi has a "-1 hour" participation, but her check-in time is 16:17 and her check-out time is 19:00, which should be 2.72 hours.
- In the screenshot below, I changed Yi's check-out time to 13:00, but the hours participated displayed a positive number 3.285.
image

**Comment by ShmulTomer** [comment]
> Overall it looks fantastic! I am so glad that this finally worked out!!!
> 
> There seem to be some issues with the calculation of participated hours. For example,
> 
> * In steps [#94](https://go.playerzero.app/share/63decbb12b1d69c0be54997a/devtools/94/Comments) to [#97](https://go.playerzero.app/share/63decbb12b1d69c0be54997a/devtools/97/Comments) in this report: (https://go.playerzero.app/share/63decbb12b1d69c0be54997a), there is a Yi has a "-1 hour" participation, but her check-in time is 16:17 and her check-out time is 19:00, which should be 2.72 hours.
> * In the screenshot below, I changed Yi's check-out time to 13:00, but the hours participated displayed a positive number 3.285.
>   image

I definitely see a few issues! Sadly I can't get to it today but I can try for tomorrow -> I think the issues extend past the stats-over-time screen so I'm not sure it'll be a quick fix

**Comment by colin4554** [comment]
> Overall it looks fantastic! I am so glad that this finally worked out!!!
> 
> There seem to be some issues with the calculation of participated hours. For example,
> 
> * In steps [#94](https://go.playerzero.app/share/63decbb12b1d69c0be54997a/devtools/94/Comments) to [#97](https://go.playerzero.app/share/63decbb12b1d69c0be54997a/devtools/97/Comments) in this report: (https://go.playerzero.app/share/63decbb12b1d69c0be54997a), there is a Yi has a "-1 hour" participation, but her check-in time is 16:17 and her check-out time is 19:00, which should be 2.72 hours.
> * In the screenshot below, I changed Yi's check-out time to 13:00, but the hours participated displayed a positive number 3.285.
>   image

@YiQin93 This is an existing issue. I already have a ticket to standardize data/time and will add a note about this issue there. It looks like we could easily prevent it by only allowing the check-out time to be in the future (and the check-in time in the past) when admins edit time cards.

**Review by colin4554 — APPROVED** [review]

**Pull request merged**